File: modules/ella-terminal/android/src/main/java/com/margelo/nitro/ella/terminal/EllaTerminalOutputPump.kt
package com.margelo.nitro.ella.terminal

import android.os.Handler
import java.util.ArrayDeque

internal class EllaTerminalOutputPump(
  private val handler: Handler,
  private val onFeed: (EllaSshConnection, ByteArray) -> Unit,
  private val onOverflow: (EllaSshConnection) -> Unit,
) {
  private val lock = Any()
  private val chunks = ArrayDeque<OutputChunk>()
  private var queuedBytes = 0
  private var drainScheduled = false
  private var active = true
  private var overflowedConnection: EllaSshConnection? = null

  fun enqueue(connection: EllaSshConnection, bytes: ByteArray) {
    if (bytes.isEmpty()) return
    val shouldSchedule: Boolean
    synchronized(lock) {
      if (!active || overflowedConnection === connection) return
      val last = chunks.peekLast()
      val canCoalesce = last?.connection === connection &&
        last.bytes.size + bytes.size <= MAXIMUM_CHUNK_BYTES
      if (
        queuedBytes + bytes.size > MAXIMUM_QUEUED_BYTES ||
        (!canCoalesce && chunks.size >= MAXIMUM_QUEUED_CHUNKS)
      ) {
        chunks.clear()
        queuedBytes = 0
        overflowedConnection = connection
        handler.post { onOverflow(connection) }
        return
      }
      if (canCoalesce) {
        val combined = ByteArray(last.bytes.size + bytes.size)
        last.bytes.copyInto(combined)
        bytes.copyInto(combined, destinationOffset = last.bytes.size)
        chunks.removeLast()
        chunks.addLast(OutputChunk(connection, combined))
      } else {
        chunks.addLast(OutputChunk(connection, bytes.copyOf()))
      }
      queuedBytes += bytes.size
      shouldSchedule = !drainScheduled
      drainScheduled = true
    }
    if (shouldSchedule) handler.post(::drain)
  }

  fun reset() {
    synchronized(lock) {
      if (!active) return
      chunks.clear()
      queuedBytes = 0
      drainScheduled = false
      overflowedConnection = null
    }
  }

  fun invalidate() {
    synchronized(lock) {
      active = false
      chunks.clear()
      queuedBytes = 0
      drainScheduled = false
      overflowedConnection = null
    }
  }

  private fun drain() {
    val batch = mutableListOf<OutputChunk>()
    synchronized(lock) {
      if (!active) {
        drainScheduled = false
        return
      }
      var batchBytes = 0
      while (chunks.isNotEmpty() && batchBytes < MAXIMUM_DRAIN_BYTES) {
        val chunk = chunks.removeFirst()
        queuedBytes -= chunk.bytes.size
        batchBytes += chunk.bytes.size
        batch.add(chunk)
      }
    }

    batch.forEach { onFeed(it.connection, it.bytes) }
    val shouldReschedule = synchronized(lock) {
      if (active && chunks.isNotEmpty()) {
        true
      } else {
        drainScheduled = false
        false
      }
    }
    if (shouldReschedule) handler.post(::drain)
  }

  private data class OutputChunk(
    val connection: EllaSshConnection,
    val bytes: ByteArray,
  )

  private companion object {
    const val MAXIMUM_QUEUED_BYTES = 2 * 1024 * 1024
    const val MAXIMUM_QUEUED_CHUNKS = 512
    const val MAXIMUM_DRAIN_BYTES = 64 * 1024
    const val MAXIMUM_CHUNK_BYTES = 16 * 1024
  }
}

File: modules/ella-terminal/ios/EllaTerminalOutputPump.swift
import Foundation
import SwiftTerm

final class EllaTerminalOutputPump {
  private let lock = NSLock()
  private weak var terminalView: TerminalView?
  private var chunks: [[UInt8]] = []
  private var readIndex = 0
  private var queuedBytes = 0
  private var drainScheduled = false
  private var active = true
  private let onOverflow: () -> Void

  private static let maximumQueuedBytes = 2 * 1024 * 1024
  private static let maximumQueuedChunks = 512
  private static let maximumDrainBytes = 64 * 1024
  private static let maximumFeedBytes = 16 * 1024

  init(terminalView: TerminalView, onOverflow: @escaping () -> Void) {
    self.terminalView = terminalView
    self.onOverflow = onOverflow
  }

  func enqueue(_ bytes: [UInt8]) {
    guard !bytes.isEmpty else { return }

    lock.lock()
    guard active else {
      lock.unlock()
      return
    }
    let canCoalesce = readIndex < chunks.count &&
      chunks[chunks.count - 1].count + bytes.count <= Self.maximumFeedBytes
    guard
      queuedBytes + bytes.count <= Self.maximumQueuedBytes,
      canCoalesce || chunks.count - readIndex < Self.maximumQueuedChunks
    else {
      active = false
      chunks.removeAll(keepingCapacity: false)
      readIndex = 0
      queuedBytes = 0
      lock.unlock()
      onOverflow()
      return
    }

    if canCoalesce {
      chunks[chunks.count - 1].append(contentsOf: bytes)
    } else {
      chunks.append(bytes)
    }
    queuedBytes += bytes.count
    let shouldSchedule = !drainScheduled
    drainScheduled = true
    lock.unlock()

    if shouldSchedule {
      DispatchQueue.main.async { [weak self] in
        self?.drain()
      }
    }
  }

  func invalidate() {
    lock.lock()
    active = false
    chunks.removeAll(keepingCapacity: false)
    readIndex = 0
    queuedBytes = 0
    lock.unlock()
  }

  private func drain() {
    var batch: [[UInt8]] = []
    var batchBytes = 0

    lock.lock()
    guard active else {
      drainScheduled = false
      lock.unlock()
      return
    }
    while readIndex < chunks.count, batchBytes < Self.maximumDrainBytes {
      let chunk = chunks[readIndex]
      readIndex += 1
      queuedBytes -= chunk.count
      batch.append(chunk)
      batchBytes += chunk.count
    }
    let hasMore = readIndex < chunks.count
    if !hasMore {
      chunks.removeAll(keepingCapacity: true)
      readIndex = 0
    } else if readIndex >= 256 {
      chunks.removeFirst(readIndex)
      readIndex = 0
    }
    drainScheduled = hasMore
    lock.unlock()

    for bytes in batch {
      var offset = 0
      while offset < bytes.count {
        let end = min(offset + Self.maximumFeedBytes, bytes.count)
        terminalView?.feed(byteArray: bytes[offset..<end])
        offset = end
      }
    }

    if hasMore {
      DispatchQueue.main.async { [weak self] in
        self?.drain()
      }
    }
  }
}
